### PR TITLE
Technical/graphics api abstraction

### PR DESCRIPTION
**Pull request by mazbox**
here is the start of the API to insert a Metal/Sokol back end. There is no more GL code in the Graphics class.

### CODE REVIEW

File: lib/mzgl/gl/Graphics.h [hunk 1]
```diff
@@ -220,12 +225,10 @@ class Graphics {
 	glm::mat4 viewProjectionMatrix;
 	friend class ScopedAlphaBlend;
 	friend struct ScopedTranslate;
-
+	friend class GraphicsAPI;
+	friend class OpenGLAPI;
```
Comment (PaulChana) [inline lib/mzgl/gl/Graphics.h:229]: using `friend class` always makes me assume there is a hack somewhere - Are they really needed, or could we lean harder on making GraphicsApi a truly pure virtual interface?

File: lib/mzgl/gl/Graphics.cpp [hunk 2]
```diff
@@ -428,7 +222,7 @@ float Graphics::getStrokeWeight() {
 void Graphics::setStrokeWeight(float f) {
 	if (f != this->strokeWeight) {
 		this->strokeWeight = f;
```
Comment (PaulChana) [inline lib/mzgl/gl/Graphics.cpp:224]: I used to write `this->` everywhere too - but someone pointed out to me about how easy it is to make errors when refactoring code when using this pointers, so I stopped doing it. Now I find it a bit visually jarring (Obviously this wasn't introduced on this PR, just commented on it because it stood out)

File: lib/mzgl/gl/Graphics.cpp [hunk 3]
```diff
@@ -527,12 +321,11 @@ void Graphics::drawCircle(glm::vec2 c, float r) {
 	drawCircle(c.x, c.y, r);
 }
 void Graphics::drawCircle(float x, float y, float r) {
-	vector<glm::vec2> verts;
-	//vector<glm::vec4> cols;
+	std::vector<glm::vec2> verts;
 
 	int circleResolution = 100;
```
Comment (PaulChana) [inline lib/mzgl/gl/Graphics.cpp:326]:
```suggestion
	static constexpr auto circleResolution = 100;
```

File: lib/mzgl/gl/Graphics.cpp [hunk 4]
```diff
@@ -527,12 +321,11 @@ void Graphics::drawCircle(glm::vec2 c, float r) {
...
 	verts.reserve(circleResolution + 2);
-	//verts.push_back(glm::vec2(x, y));
+
 	for (int i = 0; i <= circleResolution; i++) {
 		float phi = M_PI * 2.f * i / (float) circleResolution;
 		verts.emplace_back(x + cos(phi) * r, y + sin(phi) * r);
```
Comment (PaulChana) [inline lib/mzgl/gl/Graphics.cpp:331]: In fact, if `circleResolution` was a constexpr, you could then build a LUT of `phi` and avoid that computation. You could build an static constexpr array of std::pair(cos(phi), sin(phi)) and then this function would become 2 adds, 2 multiply and an emplace_back

File: lib/mzgl/gl/Graphics.cpp [hunk 5]
```diff
@@ -572,140 +365,34 @@ void Graphics::drawArc(glm::vec2 c, float r, float startAngle, float endAngle) {
...
 }
 
 void Graphics::maskOff() {
-	glDisable(GL_SCISSOR_TEST);
+	api->maskOff();
 }
 
 bool Graphics::isMaskOn() {
-	return glIsEnabled(GL_SCISSOR_TEST);
+	return api->isMaskOn();
```
Comment (PaulChana) [inline lib/mzgl/gl/Graphics.cpp:391]: This just made me think of :

gif

LOL

File: lib/mzgl/gl/Graphics.cpp [hunk 6]
```diff
@@ -760,7 +445,7 @@ void Graphics::drawRoundedRectShadow(Rectf r, float radius, float shadow) {
 
 #include "DefaultFont.inc.h"
 
-vector<unsigned char> Graphics::getDefaultFontTTFData() {
+std::vector<unsigned char> Graphics::getDefaultFontTTFData() {
```
Comment (PaulChana) [inline lib/mzgl/gl/Graphics.cpp:448]: This is what I was alluding to earlier. Each call here will make a memory copy of the data. If you make it a static constexpr array and the compiler can see straight through it and will give you it without the memory over head

File: lib/mzgl/gl/Graphics.cpp [hunk 7]
```diff
@@ -858,3 +541,8 @@ void Graphics::clear(float c) {
 void Graphics::clear(float r, float g, float b, float a) {
 	clear({r, g, b, a});
 }
+
+#include "OpenGLAPI.h"
+Graphics::Graphics() {
+	api = std::make_unique<OpenGLAPI>(*this);
```
Comment (PaulChana) [inline lib/mzgl/gl/Graphics.cpp:547]: Ah okay, I see the need for the friend class now. I think a factory would be a better option - That way Graphics stays clean and the implementations are clear to behave in whatever way they need to...

File: lib/mzgl/gl/Vbo.cpp [hunk 8]
```diff
@@ -12,13 +12,10 @@
...
 #include "Graphics.h"
-#include "maths.h"
 #include "log.h"
 #include "Graphics.h"
 #include "util.h"
 #include "Geometry.h"
-#include "mzAssert.h"
-#include "mainThread.h"
 
 using namespace std;
```
Comment (PaulChana) [inline lib/mzgl/gl/Vbo.cpp:20]: 💩

File: lib/mzgl/gl/api/GraphicsAPI.h [hunk 9]
```diff
@@ -0,0 +1,28 @@
+
+#pragma once
+
+class GraphicsAPI {
+public:
+	GraphicsAPI(Graphics &g)
+		: g(g) {}
+	Graphics &g;
```
Comment (PaulChana) [inline lib/mzgl/gl/api/GraphicsAPI.h:8]: Should be `protected`

File: lib/mzgl/gl/api/GraphicsAPI.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+	virtual void init()										 = 0;
+	virtual void setBlending(bool shouldBlend)				 = 0;
+	virtual void setLineWidth(float f)						 = 0;
+	virtual void setBlendMode(Graphics::BlendMode blendMode) = 0;
+	virtual void clear(vec4 c)								 = 0;
+
+	virtual void maskOn(const Rectf &r) = 0;
+	virtual void maskOff()				= 0;
+	virtual bool isMaskOn()				= 0;
+	virtual Rectf getMaskRect()			= 0;
```
Comment (PaulChana) [inline lib/mzgl/gl/api/GraphicsAPI.h:19]:
```suggestion
	[[nodiscard]] virtual bool isMaskOn() const				= 0;
	[[nodiscard]] virtual Rectf getMaskRect() const			= 0;
```

File: lib/mzgl/gl/api/GraphicsAPI.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+	virtual void readScreenPixels(uint8_t *data, const Rectf &r) = 0;
```
Comment (PaulChana) [inline lib/mzgl/gl/api/GraphicsAPI.h:21]: This function definition seems *very* dangerous. The problems that I can see are:
* data not big enough
* data way too big (how does the user know how much data to pass?)
* no size parameter passed with data to tell the function what the upper limit is
I think I would prefer to see it do something like this:
```
[[nodiscard]] virtual std::vector<uint8_t> readScreenPixels(const Rectf &bounds) =0;
```
or similar

File: lib/mzgl/gl/api/OpenGLAPI.h [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+			Log::d() << "OpenGL Version: " << openglVersion;
+#ifdef MZGL_GL2
+			Log::d() << "MZGL Compiled with MZGL_GL2==TRUE";
+#else
+			Log::d() << "MZGL Compiled with MZGL_GL2==FALSE";
+#endif
+		} else {
+			Log::d() << "OpenGL Version: null";
+		}
+		loadDefaultShaders();
```
Comment (PaulChana) [inline lib/mzgl/gl/api/OpenGLAPI.h:25]: Could all this code go in a .cpp file? It makes the interfaces much easier to read and understand

File: lib/mzgl/gl/api/OpenGLAPI.h [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+			in lowp vec4 colorV;
+			in lowp vec2 texCoordV;
+			out vec4 fragColor;
+			uniform sampler2D myTextureSampler;
+
+			void main(void) {
+				fragColor = texture(myTextureSampler, texCoordV) * colorV;
+			}
+
+		));
```
Comment (PaulChana) [inline lib/mzgl/gl/api/OpenGLAPI.h:137]: One way ive done this before is to have the strings in a separate file, and then just #include them in the code here. that way the shader code doesn't get obscured by the actual shader strings. At the moment the actual function and its behaviour is a bit hard to follow because of the the STRINGIFY parts

File: lib/mzgl/gl/api/OpenGLAPI.h [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+
+		g.currShader->uniform("mvp", g.getMVP());
+		if (g.currShader->needsColorUniform) {
+			g.currShader->uniform("color", g.getColor());
+		}
+
+#ifndef MZGL_GL2
+		if (immediateVertexArray == 0) return;
+		glBindVertexArray(immediateVertexArray);
+#endif
```
Comment (PaulChana) [inline lib/mzgl/gl/api/OpenGLAPI.h:229]: What mode do we use? I wonder if we should drop the other one in the cause of readability / maintainability

File: lib/mzgl/gl/api/OpenGLAPI.h [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+
+#ifndef MZGL_GL2
+		glBindVertexArray(0);
+#endif
+	}
+
+	void drawVerts(const std::vector<glm::vec2> &verts,
+				   const std::vector<glm::vec4> &cols,
+				   Vbo::PrimitiveType type) override {
+		g.colorShader->begin();
```
Comment (PaulChana) [inline lib/mzgl/gl/api/OpenGLAPI.h:250]: Im not a big fan of single letter variable names. I assume this is the graphics class here?

File: lib/mzgl/gl/api/OpenGLAPI.h [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+		switch (blendMode) {
+			case Graphics::BlendMode::Alpha: glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA); break;
+			case Graphics::BlendMode::Additive: glBlendFunc(GL_SRC_ALPHA, GL_ONE); break;
+			case Graphics::BlendMode::Multiply: glBlendFunc(GL_DST_COLOR, GL_ONE_MINUS_SRC_ALPHA); break;
+			case Graphics::BlendMode::Screen: glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_COLOR); break;
+			case Graphics::BlendMode::Subtract: glBlendFunc(GL_ZERO, GL_ONE_MINUS_SRC_COLOR); break;
+		}
+	}
+
+	static int primitiveTypeToGLMode(Vbo::PrimitiveType mode) {
```
Comment (PaulChana) [inline lib/mzgl/gl/api/OpenGLAPI.h:316]: Could this be private?

File: lib/mzgl/gl/Graphics.h [hunk 17]
```diff
@@ -73,6 +76,9 @@ class Graphics {
 	enum class BlendMode {
 		Alpha, // the classic alpha blending mode
```
[the remark this answers is not in the document]
Comment (mazbox) [reply lib/mzgl/gl/Graphics.h:77]: sorry, think that was a chatgpt fart

File: lib/mzgl/gl/api/OpenGLAPI.h (hunk 13 above)
Comment (mazbox) [reply lib/mzgl/gl/api/OpenGLAPI.h:137]: good point - this will all have to change with the new shader api, but for now I will sling these in a separate file in any case

File: lib/mzgl/gl/api/OpenGLAPI.h (hunk 14 above)
Comment (mazbox) [reply lib/mzgl/gl/api/OpenGLAPI.h:229]: one day we will drop GL2 but its the default at the moment - once the Sokol stuff works and we know it works on older Macs I'm happy to relinquish MZGL_GL2

File: lib/mzgl/gl/api/OpenGLAPI.h (hunk 16 above)
Comment (mazbox) [reply lib/mzgl/gl/api/OpenGLAPI.h:316]: yes, its now static and in the cpp file

File: lib/mzgl/gl/Graphics.h [hunk 18]
```diff
@@ -73,6 +76,9 @@ class Graphics {
...
 		Additive,
+		Multiply,
+		Screen,
+		Subtract,
```
[the remark this answers is not in the document]
Comment (mazbox) [reply lib/mzgl/gl/Graphics.h:81]: we're mostly using UpperCamelCase for enum classes - see Vbo::PrimitiveMode and the SP404 stuff - maybe it would be nicer for them to be lowerCamelCase - but it would be good to do that on a pass of the develop branch

File: lib/mzgl/gl/Graphics.cpp [hunk 19]
```diff
@@ -15,19 +15,19 @@
...
-using namespace std;
+#include "GraphicsAPI.h"
 
+Graphics::~Graphics() = default;
 glm::vec4 hexColor(int hex, float a) {
 	glm::vec4 c;
 	c.r = ((hex >> 16) & 0xFF) / 255.f; // Extract the RR byte
 	c.g = ((hex >> 8) & 0xFF) / 255.f; // Extract the GG byte
-	c.b = ((hex) &0xFF) / 255.f; // Extract the BB byte
+	c.b = ((hex) & 0xFF) / 255.f; // Extract the BB byte
```
[the remark this answers is not in the document]
Comment (mazbox) [reply lib/mzgl/gl/Graphics.cpp:30]: haha another chatgpt generated function. deleted

File: lib/mzgl/gl/Graphics.h (hunk 1 above)
Comment (mazbox) [reply lib/mzgl/gl/Graphics.h:229]: got rid of it now cheers, good point, didn't like it either

File: lib/mzgl/gl/Graphics.cpp (hunk 7 above)
Comment (mazbox) [reply lib/mzgl/gl/Graphics.cpp:547]: I only needed the friend thing for one private variable I wanted to share - it was the defaultFBO, I've fixed it with a dynamic_cast (may also be a bit hacky)

File: lib/mzgl/gl/Graphics.cpp (hunk 2 above)
Comment (mazbox) [reply lib/mzgl/gl/Graphics.cpp:220]: I hate this-> too - will fix here

File: lib/mzgl/gl/api/GraphicsAPI.h (hunk 11 above)
Comment (mazbox) [reply lib/mzgl/gl/api/GraphicsAPI.h:23]: ah yes, that would be a nice way to do it - this function isn't used in koala - it was used for screenshotting for debugging stuff - the only thin I would argue it to take out data as a parameter rather than a RVO'd return is that you are likely to be calling this multiple times so may want to reuse the vector - but will c++erize it with a std::vector instead of the raw pointer

File: lib/mzgl/gl/DefaultFont.inc.h [hunk 20]
```diff
@@ -7,8 +7,8 @@
 //
 // generated from this command: xxd -i ProggySquare.ttf
 
-vector<unsigned char> getDefaultFontData() {
-	vector<unsigned char> defaultFontData = {
+std::vector<unsigned char> getDefaultFontData() {
+	std::vector<unsigned char> defaultFontData = {
```
[the remark this answers is not in the document]
Comment (mazbox) [reply lib/mzgl/gl/DefaultFont.inc.h:11]: I agree but once its an array how do I pass to a method that only takes a vector? - (I think this only gets used if you turn on debug text in koala and it will happen once)

File: lib/mzgl/gl/Graphics.cpp (hunk 4 above)
Comment (mazbox) [reply lib/mzgl/gl/Graphics.cpp:327]: apparently trig functions aren't constexpr-able - but I can make it static so at least it only happens once

**Comment by mazbox** [comment]
> Happy to go with upper camel case if thats the preference? Doesn’t bother me either way as long as we are consistent :)
For now yes, I've tried to get most of the ALL_CAPS stuff out

File: lib/mzgl/gl/api/OpenGLAPI.h (hunk 15 above)
Comment (mazbox) [reply lib/mzgl/gl/api/OpenGLAPI.h]: I try to not have too many but Graphics is just used so much (as well as R) that I let those ones slide.
